torch/testing/_internal/common_utils.py

```diff
@@ -1049,6 +1049,32 @@ def setUp(self):
 
         set_rng_seed(SEED)
 
+    def genSparseCSRTensor(self, size, nnz, *, device, dtype, index_dtype):
+        sparse_dim = 2
+        assert all(size[d] > 0 for d in range(sparse_dim)) or nnz == 0, 'invalid arguments'
+        assert len(size) == sparse_dim
+
+        def random_sparse_csr(n_rows, n_cols, nnz):
+            nnz_per_row = nnz // n_rows
+            if nnz_per_row > 0:
+                crow_indices = torch.zeros(n_rows + 1, dtype=index_dtype)
+                crow_indices[1:] = nnz_per_row
+                crow_indices.cumsum_(dim=0)
+                col_indices = torch.randint(0, n_cols, size=[nnz_per_row * n_rows], dtype=index_dtype, device=device)
+            else:
+                crow_indices = torch.zeros(n_rows + 1, dtype=index_dtype)
+                crow_indices[1:nnz + 1] = nnz_per_row
```

Comment:
This is a bug: `nnz_per_row` is 0 here and the result will be all zero `crow_indices`. I think what you meant is
```
crow_indices[1:nnz + 1] = 1
```

In fact, this reveals a bug in `torch.sparse_csr_tensor`: the invalidation of invariant `crow_indices[-1] == nnz` is not catched.

Reply:
The sparse_csr_tensor bug is reported: https://github.com/pytorch/pytorch/issues/56856

Reply:
Thanks for the feedback @pearu, I fixed the `random_sparse_csr` function. Related the invariant, yes there was not a check when the size parameter was not provided to the factory function, so I added the check, however I think that the check is `crow_indices[-1] <= nnz`  as it is described here: 

https://github.com/pytorch/pytorch/blob/267b554b6fd7d8dfdfb9885c663fbe48fb26b2f8/aten/src/ATen/native/sparse/SparseCsrTensor.cpp#L58

To support this I tested matmul operation with `scipy/sparse` and csr `matvec` pytorch implementation with mkl, both accept tensors with this kind of inputs. 
 
See test here:
https://gist.github.com/aocsa/ecdbfdc5e8d76c1703080076d1065aaa

Reply:
The check `crow_indices[-1] <= nnz` is wrong but let's resolve it in the context of issue #56856.

But in short, `crow_indices` must be increasing sequence and `diff(crow_indices)` gives the number of elements of all rows. That is, we have an invariant `sum(diff(crow_indices)) == nnz`. If you expand this expression, you'll get
```
crow_indices[1] - crow_indices[0] + crow_indices[2] - crow_indices[1] + ... + crow_indices[-1] - crow_indices[-2] == nnz
```
or after reduction:
```
crow_indices[-1] - crow_indices[0] == nnz
```
Since `crow_indices[0] = 0` by definition, this proves that `crow_indices[-1]` must always be equal to `nnz`.

This invariant is best tested for the case where `nnz < size[0]`. This case likely originates the heuristic condition `crow_indices[-1] <= nnz` but it is wrong because the content of `crow_indices` is left incomplete. To complete it, one must apply:
```
for n in range(k, size[0] + 1):
        crow_indices[n] = nnz
```